trieste/acquisition/optimizer.py

```diff
@@ -11,42 +11,47 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+r"""
+This module contains functionality for optimizing
+:data:`~trieste.acquisition.AcquisitionFunction`\ s over :class:`~trieste.space.SearchSpace`\ s.
+"""
+from __future__ import annotations
+
 from functools import singledispatch
-from typing import Callable
+from typing import Callable, TypeVar
 
 import gpflow
 import tensorflow as tf
 import tensorflow_probability as tfp
 
 from ..space import Box, DiscreteSearchSpace, SearchSpace
 from ..type import TensorType
+from .function import AcquisitionFunction
 
-TensorMapping = Callable[[TensorType], TensorType]
-
+SP = TypeVar("SP", bound=SearchSpace)
+""" Type variable bound to :class:`~trieste.space.SearchSpace`. """
 
-@singledispatch
-def optimize(space: SearchSpace, target_func: TensorMapping) -> TensorType:
-    """
-    Return the point in ``space`` (with shape S) that maximises the function ``target_func``, as the
-    single entry in a 1 by S tensor.
 
-    ``target_func`` must satisfy the following:
+AcquisitionOptimizer = Callable[[SP, AcquisitionFunction], TensorType]
+"""
+Type alias for a function that returns the single point that maximizes an acquisition function over
+a search space. For a search space with points of shape S, and acquisition function with input shape
+[...] + S output shape [..., 1], the :data:`AcquisitionOptimizer` return shape should be [1] + S.
+"""
 
-      * indices in the leading dimension for both the argument and the result of ``target_func``
-        must run over points in the space. For example, the element at index 0 is the first point,
-        and the element at index 1 is the second.
-      * the result of ``target_func`` must have exactly one additional dimension of size 1. This is
-        needed to unambiguously define a maximum.
 
-    :param space: The space of points over which to search.
-    :param target_func: The function to maximise.
-    :return: The point in ``space`` that maximises ``target_func``.
+@singledispatch
+def optimize(space: Box | DiscreteSearchSpace, target_func: AcquisitionFunction) -> TensorType:
+    """
+    :param space: The space of points over which to search, for points with shape [D].
+    :param target_func: The function to maximise, with input shape [..., D] and output shape
+        [..., 1].
+    :return: The point in ``space`` that maximises ``target_func``, with shape [1, D].
     """
-    raise TypeError(f"No optimize implementation found for space of type {type(space)}")
 
```

Comment:
This is great for future compatibility!

Reply:
great! any bit in particular?

Reply:
For when I add in spaces of strings and we have a genetic algorithm that supports maximisation over the string space, for example.

Reply:
Basically, we can plug and play in new spaces and their optimisers